apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs

```diff
@@ -1520,78 +1568,70 @@ void RetryAcksAndExtends(bool? ack, AckError ackError, IEnumerable<TimedId> exte
-                        return;
+                        // Update back-off of retryableErrorIds in _retryableIds dictionary.
+                        _retryableIds[id] = retryInfo.WithBackoff(backoff);
                     }
 
-                    // If we have ids to retry and we are not past the specified timeout, then retry.
-                    // Update backoff of retryableErrorIds in _retryableIds dictionary.
-                    UpdateAllRetryableErrorBackoff(retryableErrorIds, backoff);
-                    // Retry errors that are still temporary errors. They will be retried after backOff.
-                    Task delayTask = _scheduler.Delay(backoff, _softStopCts.Token);
-                    // TODO: Check, if the below expression can be simplified or optimized.
-                    Add(delayTask, new NextAction(false, hasAcksOrNacks
-                    ? () => { ackActionToRetry(retryableErrorIds); StartPush(); }
-                    : () => { extendActionToRetry(extendIds.Where(j => retryableErrorIds.Contains(j.Id))); StartPush(); }));
-                }
-
-                // Updates the backoff of the retryableErrorIds in _retryableIds dictionary.
-                void UpdateAllRetryableErrorBackoff(IEnumerable<string> ids, TimeSpan? backoff)
-                {
-                    foreach (var id in ids)
+                    // The ids that exist in both retryableErrorIds and _retryableIds now are the ones that need to be retried.
+                    // We can have ids with different back-off.
+                    // Group these ids on back-off and add them to the corresponding queue in the increasing order of backoff
+                    foreach (var retryGroup in _retryableIds.Where(pair => retryableErrorIds.Contains(pair.Key)).GroupBy(pair => pair.Value.Backoff).OrderBy(pair => pair.Key))
                     {
-                        if (_retryableIds.TryGetValue(id, out var info))
-                        {
-                            _retryableIds[id] = info.WithBackoff(backoff);
-                        }
+                        var backoff = retryGroup.Key ?? TimeSpan.Zero;
+                        var retryIds = retryGroup.Select(j => j.Key);
+                        Task delayTask = _scheduler.Delay(backoff, _softStopCts.Token);
+                        Add(delayTask, new NextAction(false, hasAcksOrNacks
+                        ? () => { ackActionToRetry(retryIds); StartPush(); }
+                        : () => { extendActionToRetry(extendIds.Where(j => retryIds.Contains(j.Id))); StartPush(); }));
                     }
                 }
 
-                // Updates the _retryableIds dictionary. Ensures that the ids that should be retried exist in the dictionary.
-                // By design, only ids that are in _retryableIds dictionary will be retried with exponential backoff.
+                // Updates the _retryableIds dictionary. Ensures that the ids that should be re-tried exist in the dictionary.
```

Comment:
I don't think there's any need for this change. (And retryable is fine as one word as well, without hyphenation - ditto backoff usually.)
If you find it easier to read this way though, that's fine. You're more likely to be maintaining it than I am, and it's not in documentation.

Reply:
Sure Jon. Updated the comments. Removed hyphen.